File: 프로그래머스/2/81302. 거리두기 확인하기/거리두기 확인하기.java
import java.util.*;

class Solution {
    
    static int dx[] = {-1, 0, 1, 0}, dy[] = {0, -1, 0, 1};
    
    public int[] solution(String[][] places) {
        
        int N = places.length;
        int answer[] = new int[N];
        
        for(int i=0; i<N; i++){
            String[][] wRoom = new String[5][5];
            for(int j=0; j<5; j++){
                wRoom[j] = places[i][j].split("");
            }
            
            boolean isPoss = checkDis(wRoom);
            
            if(isPoss) answer[i] = 1;
            
        }
      
        return answer;
    }
    
    public boolean checkDis(String[][] wRoom){
        
        for(int i=0; i<5; i++){
            for(int j=0; j<5; j++){
                if(wRoom[i][j].equals("P")){
                    
                    Deque<int[]> que = new ArrayDeque<>();
                    que.add(new int[]{i, j, 0});
                    boolean visited[][] = new boolean[5][5];
                    visited[i][j] = true;
                    
                    while(!que.isEmpty()){
                        int[] now = que.poll();
                        if(now[2] >= 2) continue;
                        for(int k=0; k<4; k++){
                            int nextX = now[0] + dx[k];
                            int nextY = now[1] + dy[k];
                            if(nextX < 0 || nextX >= 5 || nextY < 0 || nextY >= 5) continue;
                            if(wRoom[nextX][nextY].equals("X")) continue;
                            if(visited[nextX][nextY]) continue;
                            visited[nextX][nextY] = true;
                            if(wRoom[nextX][nextY].equals("P") && now[2] + 1 <= 2) {
                                return false;
                            }
                            que.add(new int[]{nextX, nextY, now[2]+1});
                        }
                    }
                }
            }
        }
        
        return true;
        
    }
    
}

File: 프로그래머스/2/81302. 거리두기 확인하기/거리두기 확인하기.js
const dx = [0, -1, 0, 1]
const dy = [1, 0, -1, 0]

function solution(places) {
    
    var answer = Array(5).fill(0);
    
    for(let i=0; i<5; i++){
        var wRoom = Array.from({length : 5}, ()=> Array(5).fill(0))
        for(let j=0; j<5;j++){
            wRoom[j] = places[i][j].split("");
        }
        var isPoss = checkDis(wRoom);
        if(isPoss) answer[i] = 1;
    }
    
    return answer;
}

function checkDis(wRoom){
    
    for(let i=0; i<5; i++){
        for(let j=0; j<5; j++){
            if(wRoom[i][j] == "P"){
                const que = [];
                que.push([i, j, 0]);
                const visited = Array.from({length : 5}, () => Array(5).fill(false));
                visited[i][j] = true;
                while(que.length){
                    var now = que.shift();
                    if(now[2] >= 2) continue;
                    for(let k=0; k<4; k++){
                        var nextX = now[0] + dx[k];
                        var nextY = now[1] + dy[k];
                        if(nextX < 0 || nextX >= 5 || nextY < 0 || nextY >= 5) continue;
                        if(visited[nextX][nextY]) continue;
                        visited[nextX][nextY] = true;
                        if(wRoom[nextX][nextY] == "X") continue;
                        if(wRoom[nextX][nextY] == "P" && now[2] + 1 <= 2){
                            return false;
                        }
                        que.push([nextX, nextY, now[2]+1])
                    }
                }
            }
        }
    }
    
    return true;
}